### PR TITLE
Barcode computer tgui port

### PR DESCRIPTION
**Pull request by TobleroneSwordfish**
[TGUI][QOL]
## About the PR 
Ports the barcode computer to tgui.

barcode_tgui
V2, slightly wider edition:
image

I've added a little display for the scanned ID card to show who is getting rich and by how much, and changed the trader and requisition options to display their names instead of codes.
Criticism appreciated as I am not too familiar with tgui yet.

## Why's this needed? 
The old UI was ugly and HTML

## Changelog

```changelog
(u)LeahTheTech
(*)The barcode computer now has a new UI, this one shows the scanned ID card clearly so you can keep track of how rich the other QM is getting.
```

### CODE REVIEW

**Comment by DisturbHerb** [comment]
My own two cents, but the ID card display doesn't match the convention used for other devices which use ID cards in their operation (such as the slot machine) and feels a bit out of place. I also just kinda dislike the way the buttons for the traders and other labels are laid out, but I can't think of a good way to replace it.

**Comment by TobleroneSwordfish** [comment]
> My own two cents, but the ID card display doesn't match the convention used for other devices which use ID cards in their operation (such as the slot machine) and feels a bit out of place. I also just kinda dislike the way the buttons for the traders and other labels are laid out, but I can't think of a good way to replace it.

I was trying to think of other tguis that use IDs and forgot slot machines existed, makes sense to standardize with those so I'll work on that. (although I do like how the ID card bit looks here so I might save that for the ID modification terminal at some point)

The labels do look weird, but I tried them in columns and that looked worse due to how long some of the names are. Maybe it'd work if I make the buttons fixed width.

File: code/obj/machinery/launcherloader.dm [hunk 1]
```diff
@@ -411,37 +431,19 @@
...
+		for (var/datum/trader/T in shippingmarket.active_traders)
+			if (T.hidden)
+				continue
+			traders += list(list("crate_tag" = T.crate_tag, "name" = T.name))
+		.["sections"] += list(list("title" = "Traders", "destinations" = traders))
+		var/list/req_codes = new()
+		req_codes += list(list("crate_tag" = "REQ-THIRDPARTY", "name" = "Third party"))
+		for (var/datum/req_contract/RC in shippingmarket.req_contracts)
+			req_codes += list(list("crate_tag" = RC.req_code, "name" = RC.name))
+		.["sections"] += list(list("title" = "Requisition contracts", "destinations" = req_codes))
```
Comment (ZeWaka) [inline code/obj/machinery/launcherloader.dm:446]: I feel like the information in here could be static data, and then manually refreshed when the active traders/contracts list changes?

File: code/obj/machinery/launcherloader.dm [hunk 2]
```diff
@@ -329,21 +330,78 @@
...
+			. = TRUE
+		else if (action == "reset_id")
+			scan = null
+			account = null
+			. = TRUE
+			src.updateUsrDialog()
+		// cogwerks - uncomment this stuff if/when custom locations are ready
+		/*else if (action == "remove")
+			if(destinations.Find(params["remove"]))
+				destinations.Remove(params["remove"])
```
Comment (ZeWaka) [inline code/obj/machinery/launcherloader.dm:399]: prob just delete this

File: code/obj/machinery/launcherloader.dm [hunk 3]
```diff
@@ -329,21 +330,78 @@
...
-		if (..(user))
+	proc/print(var/destination)
+		if (printing)
 			return
+		printing = 1
+		playsound(src.loc, "sound/machines/printer_cargo.ogg", 75, 0)
+		sleep(1.75 SECONDS)
+		for (var/i = 0; i < src.print_amount; i++)
+			var/obj/item/sticker/barcode/B = new/obj/item/sticker/barcode(src.loc)
+			// var/dest = strip_html(href_list["print"], 64)
```
Comment (ZeWaka) [inline code/obj/machinery/launcherloader.dm:341]: .

File: code/obj/machinery/launcherloader.dm [hunk 4]
```diff
@@ -329,21 +330,78 @@
...
 
-	attack_hand(var/mob/user as mob)
-		if (..(user))
+	proc/print(var/destination)
+		if (printing)
 			return
+		printing = 1
+		playsound(src.loc, "sound/machines/printer_cargo.ogg", 75, 0)
+		sleep(1.75 SECONDS)
+		for (var/i = 0; i < src.print_amount; i++)
```
Comment (ZeWaka) [inline code/obj/machinery/launcherloader.dm:339]:
```suggestion
		for (var/i in 0 to src.print_amount)
```

File: code/obj/machinery/launcherloader.dm [hunk 5]
```diff
@@ -329,21 +330,78 @@
 
 	var/list/destinations = list("Airbridge", "Cafeteria", "EVA", "Engine", "Disposals", "QM", "Catering", "MedSci", "Security") //These have to match the ones on the cargo routers for the routers to work.
 
-	attack_hand(var/mob/user as mob)
-		if (..(user))
+	proc/print(var/destination)
+		if (printing)
 			return
+		printing = 1
```
Comment (ZeWaka) [inline code/obj/machinery/launcherloader.dm:336]:
```suggestion
		printing = TRUE
```

File: code/obj/machinery/launcherloader.dm [hunk 6]
```diff
@@ -317,6 +317,7 @@
 
 	icon = 'icons/obj/delivery.dmi'
 	icon_state = "barcode_comp"
+	flags = TGUI_INTERACTIVE
 	deconstruct_flags = DECON_SCREWDRIVER | DECON_WRENCH | DECON_WIRECUTTERS | DECON_MULTITOOL
 
 	var/printing = 0
```
Comment (ZeWaka) [inline code/obj/machinery/launcherloader.dm:323]:
```suggestion
	var/printing = FALSE
```

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/obj/machinery/launcherloader.dm (hunk 1 above)
Comment (TobleroneSwordfish) [reply code/obj/machinery/launcherloader.dm:435]: Hmm, how to detect when they change though? The only way I can think of is to set a static flag on the barcode computer class from the shipping market whenever it modifies the lists and then get ui_data to check the flag and call update_static_data. This seems kinda flimsy and jank, is there a better way?

Comment (ZeWaka) [reply code/obj/machinery/launcherloader.dm:436]: Hmm not sure at the moment, can't really look into it. However, doing this data computation every 0.9s is kind of a lot.

Comment (TobleroneSwordfish) [reply code/obj/machinery/launcherloader.dm:436]: I'll try and figure out something reasonable tomorrow

Comment (ZeWaka) [reply code/obj/machinery/launcherloader.dm:436]: Why not just 
```
modify_shipping_data() / modify_trader_actives()
...
for (barcode computer in either machine registry or by_type)
	computer.update_static_data()
```

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**